src/ExpressionEvaluator/Core/Source/ExpressionCompiler/FrameDecoder.cs

```diff
@@ -19,105 +20,189 @@ namespace Microsoft.CodeAnalysis.ExpressionEvaluator
 
-        internal FrameDecoder(InstructionDecoder instructionDecoder)
+        internal FrameDecoder(InstructionDecoder<TMethodSymbol, TModuleSymbol, TTypeSymbol, TTypeParameterSymbol> instructionDecoder)
         {
             _instructionDecoder = instructionDecoder;
         }
 
-        void IDkmLanguageFrameDecoder.GetFrameName(DkmInspectionContext inspectionContext, DkmWorkList workList, DkmStackWalkFrame frame, DkmVariableInfoFlags argumentFlags, DkmCompletionRoutine<DkmGetFrameNameAsyncResult> completionRoutine)
+        void IDkmLanguageFrameDecoder.GetFrameName(
+            DkmInspectionContext inspectionContext,
+            DkmWorkList workList,
+            DkmStackWalkFrame frame,
+            DkmVariableInfoFlags argumentFlags,
+            DkmCompletionRoutine<DkmGetFrameNameAsyncResult> completionRoutine)
         {
             try
             {
                 Debug.Assert((argumentFlags & (DkmVariableInfoFlags.Names | DkmVariableInfoFlags.Types | DkmVariableInfoFlags.Values)) == argumentFlags,
                     "Unexpected argumentFlags", "argumentFlags = {0}", argumentFlags);
 
-                var instructionAddress = (DkmClrInstructionAddress)frame.InstructionAddress;
-                var includeParameterTypes = argumentFlags.Includes(DkmVariableInfoFlags.Types);
-                var includeParameterNames = argumentFlags.Includes(DkmVariableInfoFlags.Names);
-
-                if (argumentFlags.Includes(DkmVariableInfoFlags.Values))
-                {
-                    // No need to compute the Expandable bit on
-                    // argument values since that can be expensive.
-                    inspectionContext = DkmInspectionContext.Create(
-                        inspectionContext.InspectionSession,
-                        inspectionContext.RuntimeInstance,
-                        inspectionContext.Thread,
-                        inspectionContext.Timeout,
-                        inspectionContext.EvaluationFlags | DkmEvaluationFlags.NoExpansion,
-                        inspectionContext.FuncEvalFlags,
-                        inspectionContext.Radix,
-                        inspectionContext.Language,
-                        inspectionContext.ReturnValue,
-                        inspectionContext.AdditionalVisualizationData,
-                        inspectionContext.AdditionalVisualizationDataPriority,
-                        inspectionContext.ReturnValues);
+                GetNameWithGenericTypeArguments(inspectionContext, workList, frame,
+                    onSucess: method => GetFrameName(inspectionContext, workList, frame, argumentFlags, completionRoutine, method),
+                    onFailure: e => completionRoutine(DkmGetFrameNameAsyncResult.CreateErrorResult(e)));
+            }
+            catch (Exception e) when (ExpressionEvaluatorFatalError.CrashIfFailFastEnabled(e))
```

Comment:
It could, but I liked leaving this boiler-plate in the entry point since it's in all the other entry points (so it's immediately obvious by scanning the entry points that we have NFW handling).